base/services/dhcpcsvc/dhcp/pipe.c

```diff
@@ -56,10 +309,12 @@ DWORD WINAPI PipeThreadProc( LPVOID Parameter ) {
           COMM_PIPE_OUTPUT_BUFFER,
           COMM_PIPE_INPUT_BUFFER,
           COMM_PIPE_DEFAULT_TIMEOUT,
-          NULL );
+          &SecurityAttributes );
     if (CommPipe == INVALID_HANDLE_VALUE)
     {
         DbgPrint("DHCP: Could not create named pipe\n");
+        HeapFree(GetProcessHeap(), 0, DhcpPipeSD);
+        DhcpPipeSD = NULL;
```

Comment:
I wonder whether it is possible to free the `DhcpPipeSD` at once just after the `CreateNamedPipeW` call (thus avoiding the repetitive HeapFree calls below).
In principle the kernel has captured the pipe's SD and doesn't need the info (on user-mode side) to stay around anymore.
But please test first whether this suggestion doesn't break anything.

Reply:
You're correct, as once Se of the kernel has captured the descriptor and Ob assigning it to a securable object the information allocated in user mode is pretty much useless afterwards. As long as the allocated descriptor pointer in memory that you give to `CreateNamedPipeW` (and any other function in Win32 that accepts a security attribute with descriptor) is valid then the kernel happily captures it and you're good to go.